libr/include/r_util/r_str.h

```diff
@@ -19,6 +19,54 @@ typedef enum {
 	R_STRING_ENC_GUESS = 'g',
 } RStrEnc;
 
+// r_strf API can be used for short strings that are needed for a short period
+// of time. They are allocated on the stack as needed, by making sure the total
+// size of allocated strings does not become too big. If that is possible (e.g.
+// the input is not bounded and/or comes from the user) you should use the
+// malloc as usual.
+//
+// Use r_strf when you just need one string at a time. NOTE: Every time you use
+// r_strf, the previously created string can be replaced with the new one, so
+// make sure to use it only when you are done with the previous string
+//
+// Use r_strf_multi when you need to use multiple strings at a time (e.g.
+// multiple strings passed to the same function). NOTE: each string allocated
+// with r_strf_cont allocates other space on the stack, so be aware and use it
+// only for known short strings and not when inside loops as you risk of
+// allocating too much stuff.
+struct r_strf_t {
+	char *buf;
+	char *old_buf;
+	size_t sz;
+	size_t old_sz;
+	size_t total_sz;
+	int tmp;
+};
+
+#define R_STRF_MAX 2 * 1024
+#define R_STRF_FRAME r_strf_frame__
+#define r_strf_frame() struct r_strf_t R_STRF_FRAME = { 0 }
+#define r_strf(fmt, ...) ( \
+	R_STRF_FRAME.tmp = snprintf (R_STRF_FRAME.buf, R_STRF_FRAME.sz, fmt, __VA_ARGS__), \
+	R_STRF_FRAME.tmp < 0 ? r_assert_log (R_LOGLVL_FATAL, "r_strf error while using snprintf") : NULL, \
+	R_STRF_FRAME.sz = R_STRF_FRAME.tmp >= R_STRF_FRAME.sz ? R_STRF_FRAME.tmp + 1 : R_STRF_FRAME.sz, \
+	SZT_ADD_OVFCHK (R_STRF_FRAME.total_sz, R_STRF_FRAME.sz) || (R_STRF_FRAME.total_sz + R_STRF_FRAME.sz >= R_STRF_MAX) ? r_assert_log (R_LOGLVL_FATAL, "r_strf used for long strings") : NULL, \
+	R_STRF_FRAME.buf = R_STRF_FRAME.sz != R_STRF_FRAME.old_sz ? (R_STRF_FRAME.total_sz += R_STRF_FRAME.sz, alloca (R_STRF_FRAME.sz)) : R_STRF_FRAME.buf, \
+	R_STRF_FRAME.tmp = R_STRF_FRAME.sz != R_STRF_FRAME.old_sz ? snprintf (R_STRF_FRAME.buf, R_STRF_FRAME.sz, fmt, __VA_ARGS__) : R_STRF_FRAME.tmp, \
+	R_STRF_FRAME.tmp < 0 ? r_assert_log (R_LOGLVL_FATAL, "r_strf error while using snprintf") : NULL, \
+	R_STRF_FRAME.old_sz = R_STRF_FRAME.sz, \
+	R_STRF_FRAME.buf)
+
+#define r_strf_multi(fmt, ...) ( \
```

Comment:
isnt r_strf_multi rewinding to 0 and therefor destroying the previous strings allocated?

Reply:
theres no r_strf_save()  like i said

Reply:
There is no "rewind" here because I'm not using a single buffer for multiple strings... Every time you call r_strf_multi it allocates new space on the stack with alloca. When you call r_strf instead it reuses the same space if possible

Reply:
I think being able to do `sdb_set(r_strf_multi("a"), r_strf_multi("b"));` in a loop is an important use case we have to support.

Reply:
Actually yeah... this use case seems to be used in current code base and as it is right now this implementation won't work well for this. A small adjustment may be something like:

```C
r_strf_frame (n_of_strings_used_in_this_function=3);
...
char *s0 = r_strf0("%d", 32);
assert(s0 == "32")
char *s1 = r_strf1("%d", 55);
assert(s0 == "32" && s1 == "55");
char *s2 = r_strf0("%d", 22);
assert(s1 == "55" && s2 == "22" && cannot_guaranteee_s0_is_still_32);
char *s3 = r_strf4("%d", 44) // FATAL ERROR
```

so you can explicitly tell which strings are preserved at any given point and which are not. (not sure how to implement it yet, but I think it's possible). @thestr4ng3r @radare